packages/json-rpc/legacy-provider/src/upstream/proofs.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  HexString,
+  ProofTrieNode,
+  TrieNode,
+  validateProofs,
+} from "@polkadot-api/substrate-bindings"
+import { mergeMap, Observable } from "rxjs"
+
+export const createClosestDescendantMerkleValue =
+  (
+    obsRequest: <Args extends Array<any>, Payload>(
+      method: string,
+      params: Args,
+    ) => Observable<Payload>,
+  ) =>
+  (at: HexString, key: HexString) =>
+    obsRequest<
+      [keys: Array<HexString>, at: HexString],
+      {
+        at: HexString
+        proof: HexString[]
+      }
+    >("state_getReadProof", [[key], at]).pipe(
+      mergeMap((x) => {
+        const result = validateProofs(x.proof)
+        if (!result) throw new Error("Invalid Proof")
+        const { rootHash, proofs } = result
+        let winnerHash: HexString | undefined = rootHash
+        let current: {
+          hash: HexString
+          parent?: HexString
+        } & TrieNode = proofs[winnerHash!] as any
+
+        let nKeyChars = 2 // skipping `0x`
+        do {
+          const nextOne: ProofTrieNode = proofs[winnerHash!]
+          if (!nextOne || nextOne.type === "Raw") break
+
+          current = nextOne
+          winnerHash = undefined
+          if (
+            !current.partialKey.startsWith(
+              key.slice(nKeyChars, nKeyChars + current.partialKey.length),
+            )
+          )
+            return []
```

Comment:
which case is this? Won't this case end up with the `closestDescendantMerkleValue` call stalling?

Might be worth documenting it

Reply:
> which case is this?

When the the `key` belongs to an non-existing storage entry. Eg: a System.Account of an address without ED.

> Won't this case end up with the `closestDescendantMerkleValue` call stalling?

Nope. The Observable will complete, and that will trigger an [`OperatonStorageDone`](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_v1_follow.html#operationstoragedone) event without triggering the intermediary [`OperationStorageItems`](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_v1_follow.html#operationstorageitems). Which is the expected behaviour when querying a non-existing storage item.

> Might be worth documenting it

yep, for sure!